qualcommax: ipq50xx: Add support for Yuncore AX830

Shubhamvis98 opened the pull request:
qualcommax: ipq50xx: Add support for Yuncore AX830

Specifications:
SOC: Qualcomm IPQ5018 (64-bit dual-core ARM Cortex-A53 @ 1.0Ghz)
Memory: 512MB DDR3L
Standard: 802.11ax/ac/b/g/n
Flash: SPI NOR 8MB (Winbond W25Q64DW) + NAND 128MB (Winbond W25N01GWZEIG)
2.4G Frequency: 2.4GHz - 2.484GHz
2.4G Wi-Fi standard: 802.11b/g/n/ax
5.8G Frequency: 4.9~5.9G
5.8G Wi-Fi Standard: 802.11 a/n/ac/ax
Interface:
  Optional 1(Without 8081):
    1 * 10/100 /1000Mbps RJ45 WAN Port and PoE port;
    1* Gigabit Console port;
  Optional 2(With 8081):
    1 * 10/100/1000/2500Mbps RJ45 WAN port and PoE port,
    1*10/100/1000Mbps LAN port
Buttons: 1 * Reset button, press 10 seconds to revert to default setting
Antenna: Build in 4*4dBi dual band MIMO Antenna
Data Rate: 3000Mbps （2.4G 600Mbps, 5.8G 2400Mbps）
End Users: 128+
2.4G RF Power: ≤ 23dBm
5.8G RF Power: ≤ 23dBm
DC: 12V----2A
PoE: 48V (IEEE 802.3at)
LED Light: Sys, WAN, LAN
Power Consumption: ≤ 20W

BACKUP YOUR STOCK FIRMWARE:
- Put openwrt-*-initramfs-uImage.itb to your
  TFTP server and rename it to initramfs.bin
- Enable serial console and enter to u-boot cli
  and exec these commands:
    ```
    tftpboot <your_tftp_server_ip>:initramfs.bin
    bootm
    ```
- Once boot completed and you get the openwrt shell
  execute below commands:
    ```
    device=ax830
    mkdir -p /tmp/fw_backup; cd /tmp/fw_backup
    rootfs=$(cat /proc/mtd | grep \"rootfs\" | cut -d: -f1)
    rootfs_1=$(cat /proc/mtd | grep \"rootfs_1\" | cut -d: -f1)
    dd if=/dev/${rootfs} of=rootfs_${rootfs} bs=1M
    dd if=/dev/${rootfs_1} of=rootfs_1_${rootfs_1} bs=1M
    cp /sys/firmware/fdt fdt.dtb
    md5sum * > md5sum
    tar -cvzf /tmp/${device}.tar.gz .
    sum=$(md5sum /tmp/${device}.tar.gz | cut -d' ' -f1)
    mv /tmp/${device}.tar.gz /tmp/${device}_${sum}.tar.gz
    echo "stock fw backup saved to: /tmp/${device}_${sum}.tar.gz"
    ```
- Upload/save your backup to a safe place.

STOCK FIRMWARE RECOVERY:
- Boot initramfs image
- Upload your backed-up stock fw tarball to the device
  using scp or download it from the device using wget.
- Enter device ssh cli or tty and exec:
    ```
    cd /tmp && wget <your_web_server_ip>/${stock_fw_backup}.tar.gz`
    tar -xpzf ${stock_fw_backup}.tar.gz
    rootfs=$(cat /proc/mtd | grep \"rootfs\" | cut -d: -f1)
    rootfs_1=$(cat /proc/mtd | grep \"rootfs_1\" | cut -d: -f1)
    ubiformat /dev/${rootfs} -y -f /tmp/rootfs_${rootfs}
    ubiformat /dev/${rootfs_1} -y -f /tmp/rootfs_1_${rootfs_1}
    reboot
    ```

INSTALLATION:
1. initramfs method
- Put openwrt-*-initramfs-uImage.itb to your
  TFTP server and rename it to initramfs.bin
- Enable serial console and enter to u-boot cli
  and exec these commands:
    ```
    tftpboot <your_tftp_server_ip>:initramfs.bin
    bootm
    ```
- Once boot completed and you get the openwrt shell
  execute below commands:
    ```
    cd /tmp && wget <your_web_server_ip>/factory.ubi`
    export rootfs=$(cat /proc/mtd | grep rootfs | cut -d: -f1)
    export rootfs_1=$(cat /proc/mtd | grep rootfs_1 | cut -d: -f1)
    ubiformat /dev/${rootfs} -y -f factory.ubi
    ubiformat /dev/${rootfs_1} -y -f factory.ubi
    reboot
    ```

2. u-boot factory.ubi image method
- Put factory.ubi to your TFTP server and
  enter u-boot cli and exec these commands:
    ```
    tftpboot <your_tftp_server_ip>:factory.ubi
    #After downloading is finished:
    flash rootfs
    flash rootfs_1
    reset
    ```

musashino205 commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq5018-ax830.dts:
```diff
@@ -0,0 +1,491 @@
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+					label = "0:TRAINING";
+					reg = <0x00 0x80000>;
+			};
```
wrong indentations (this and other partitions)

```suggestion
			partition@0 {
				label = "0:TRAINING";
				reg = <0x00 0x80000>;
			};
```

Shubhamvis98 commented:
Thanks @musashino205, just merged your indent fix.

musashino205 commented:
Please apply the same fixes to the all partition nodes.

georgemoussalem commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq5018-ax830.dts:
```diff
@@ -0,0 +1,491 @@
+
+	pinctrl-0 = <&mdio1_pins>;
+	pinctrl-names = "default";
+	reset-gpios = <&tlmm 39 GPIO_ACTIVE_LOW>;
```
is this reset needed? it's typically the reset for the qca8337 switch which this board doesn't have.

Shubhamvis98 replied:
The reset is needed. The Wan ethernet port stops working if removed.

Shubhamvis98 replied:
Hi,
I've removed the reset-gpio property and tested again — the WAN is working now. I believe I made a mistake during my earlier testing when the WAN wasn't working. Apologies for the confusion.

Thanks to both of you for reviewing.

georgemoussalem commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq5018-ax830.dts:
```diff
@@ -0,0 +1,491 @@
+		nand-ecc-engine = <&qpic_nand>;
+
+		nand-ecc-strength = <4>;
+		nand-ecc-step-size = <512>;
```
you should be able to remove the ecc strength and step size properties, can you test? The driver now defaults to 4 bits if the ecc strength requirement is 1 (which is the case for this chip)

georgemoussalem commented on target/linux/qualcommax/image/ipq50xx.mk:
```diff
@@ -108,3 +108,18 @@ define Device/linksys_spnmx56
+	PAGESIZE := 2048
+	SOC := ipq5018
+	DEVICE_DTS_CONFIG := config@mp03.5-c1
+	DEVICE_PACKAGES := kmod-ath11k-pci \
```
no need for kmod-ath11k-pci. Both use ahb

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq5018-ax830.dts:
```diff
@@ -0,0 +1,487 @@
+
+		wlan2g {
+			color = <LED_COLOR_ID_GREEN>;
+			gpios = <&tlmm 33 GPIO_ACTIVE_LOW>;
```
Function is missing

Shubhamvis98 replied:
Added, thanks for reviewing.

georgemoussalem replied:
for 2G, add: 

`function = LED_FUNCTION_WLAN_2GHZ;` 

and

`function = LED_FUNCTION_WLAN_5GHZ;` 

for 5G

Shubhamvis98 replied:
Done, Thanks

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq5018-ax830.dts:
```diff
@@ -0,0 +1,487 @@
+		wlan2g {
+			color = <LED_COLOR_ID_GREEN>;
+			gpios = <&tlmm 33 GPIO_ACTIVE_LOW>;
+			default-status = "on";
```
Why?

Please add a trigger instead

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq5018-ax830.dts:
```diff
@@ -0,0 +1,487 @@
+	};
+
+	qpic_pins: qpic-state {
+
```
Please remove the newline

Shubhamvis98 replied:
removed as suggested.

georgemoussalem commented:
@Shubhamvis98 : pls check the test formalities. This commit misses a Signed-off-by tag

Shubhamvis98 commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq5018-ax830.dts:
```diff
@@ -0,0 +1,487 @@
+		wlan5g {
+			color = <LED_COLOR_ID_BLUE>;
+			gpios = <&tlmm 34 GPIO_ACTIVE_LOW>;
+			default-status = "on";
```
added

georgemoussalem commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq5018-ax830.dts:
```diff
@@ -0,0 +1,487 @@
+	status = "okay";
+	qcom,rproc = <&q6_wcss_pd1>;
+	qcom,ath11k-calibration-variant = "Yuncore-AX830";
+	qcom,ath11k-fw-memory-mode = <2>;
```
change fw mode to <1>

https://git.openwrt.org/?p=openwrt/openwrt.git;a=commit;h=cf715a230589058aeb703f930504d6d98619a53b

Shubhamvis98 replied:
Thanks for the reference. Just fixed.

Rondom commented:
>  As @musashino205 suggested, I tried using my real email, but the test still fails. So, is it acceptable to include two Signed-off-by lines — one with my real email and another with the masked one — to satisfy the test requirements?

Use git commit --amend --reset-author (git commit --amend --author="First last <e@mail.com>" to update the authorship info in your git commit

Shubhamvis98 commented on target/linux/qualcommax/image/ipq50xx.mk:
```diff
@@ -150,3 +150,17 @@ ifneq ($(CONFIG_TARGET_ROOTFS_INITRAMFS),)
+	PAGESIZE := 2048
+	SOC := ipq5018
+	DEVICE_DTS_CONFIG := config@mp03.5-c1
+	DEVICE_PACKAGES := ath11k-firmware-ipq5018 \
```
i've already created board files. will push to the repo soon.

Shubhamvis98 replied:
board files MR: https://github.com/openwrt/firmware_qca-wireless/pull/96

Shubhamvis98 replied:
@georgemoussalem, is there anything else I need to do to create the ipq-wifi package?

Shubhamvis98 replied:
added package ipq-wifi-yuncore_ax830

robimarko replied:
Well, you added it to default packages but it does not exist yet.

Look at ipq-wifi commits to see how you need to add your BDF-s

Shubhamvis98 committed:
qualcommax: ipq50xx: Add support for Yuncore AX830

Specifications:
SOC: Qualcomm IPQ5018 (64-bit dual-core ARM Cortex-A53 @ 1.0Ghz)
Memory: 512MB DDR3L
Standard: 802.11ax/ac/b/g/n
Flash: SPI NOR 8MB (Winbond W25Q64DW) + NAND 128MB (Winbond W25N01GWZEIG)
2.4G Frequency: 2.4GHz - 2.484GHz
2.4G Wi-Fi standard: 802.11b/g/n/ax
5.8G Frequency: 4.9~5.9G
5.8G Wi-Fi Standard: 802.11 a/n/ac/ax
Interface:
  Optional 1(Without 8081):
    1 * 10/100 /1000Mbps RJ45 WAN Port and PoE port;
    1* Gigabit Console port;
  Optional 2(With 8081):
    1 * 10/100/1000/2500Mbps RJ45 WAN port and PoE port,
    1*10/100/1000Mbps LAN port
Buttons: 1 * Reset button, press 10 seconds to revert to default setting
Antenna: Build in 4*4dBi dual band MIMO Antenna
Data Rate: 3000Mbps （2.4G 600Mbps, 5.8G 2400Mbps）
End Users: 128+
2.4G RF Power: ≤ 23dBm
5.8G RF Power: ≤ 23dBm
DC: 12V----2A
PoE: 48V (IEEE 802.3at)
LED Light: Sys, WAN, LAN
Power Consumption: ≤ 20W

BACKUP YOUR STOCK FIRMWARE:
- Put openwrt-*-initramfs-uImage.itb to your
  TFTP server and rename it to initramfs.bin
- Enable serial console and enter to u-boot cli
  and exec these commands:
    ```
    tftpboot <your_tftp_server_ip>:initramfs.bin
    bootm
    ```
- Once boot completed and you get the openwrt shell
  execute below commands:
    ```
    device=ax830
    mkdir -p /tmp/fw_backup; cd /tmp/fw_backup
    rootfs=$(cat /proc/mtd | grep \"rootfs\" | cut -d: -f1)
    rootfs_1=$(cat /proc/mtd | grep \"rootfs_1\" | cut -d: -f1)
    dd if=/dev/${rootfs} of=rootfs_${rootfs} bs=1M
    dd if=/dev/${rootfs_1} of=rootfs_1_${rootfs_1} bs=1M
    cp /sys/firmware/fdt fdt.dtb
    md5sum * > md5sum
    tar -cvzf /tmp/${device}.tar.gz .
    sum=$(md5sum /tmp/${device}.tar.gz | cut -d' ' -f1)
    mv /tmp/${device}.tar.gz /tmp/${device}_${sum}.tar.gz
    echo "stock fw backup saved to: /tmp/${device}_${sum}.tar.gz"
    ```
- Upload/save your backup to a safe place.

STOCK FIRMWARE RECOVERY:
- Boot initramfs image
- Upload your backed-up stock fw tarball to the device
  using scp or download it from the device using wget.
- Enter device ssh cli or tty and exec:
    ```
    cd /tmp && wget <your_web_server_ip>/${stock_fw_backup}.tar.gz`
    tar -xpzf ${stock_fw_backup}.tar.gz
    rootfs=$(cat /proc/mtd | grep \"rootfs\" | cut -d: -f1)
    rootfs_1=$(cat /proc/mtd | grep \"rootfs_1\" | cut -d: -f1)
    ubiformat /dev/${rootfs} -y -f /tmp/rootfs_${rootfs}
    ubiformat /dev/${rootfs_1} -y -f /tmp/rootfs_1_${rootfs_1}
    reboot
    ```

INSTALLATION:
1. initramfs method
- Put openwrt-*-initramfs-uImage.itb to your
  TFTP server and rename it to initramfs.bin
- Enable serial console and enter to u-boot cli
  and exec these commands:
    ```
    tftpboot <your_tftp_server_ip>:initramfs.bin
    bootm
    ```
- Once boot completed and you get the openwrt shell
  execute below commands:
    ```
    cd /tmp && wget <your_web_server_ip>/factory.ubi`
    export rootfs=$(cat /proc/mtd | grep rootfs | cut -d: -f1)
    export rootfs_1=$(cat /proc/mtd | grep rootfs_1 | cut -d: -f1)
    ubiformat /dev/${rootfs} -y -f factory.ubi
    ubiformat /dev/${rootfs_1} -y -f factory.ubi
    reboot
    ```

2. u-boot factory.ubi image method
- Put factory.ubi to your TFTP server and
  enter u-boot cli and exec these commands:
    ```
    tftpboot <your_tftp_server_ip>:factory.ubi
    #After downloading is finished:
    flash rootfs
    flash rootfs_1
    reset
    ```

Signed-off-by: Shubham Vishwakarma <shubhamvis98@fossfrog.in>
Link: https://github.com/openwrt/openwrt/pull/18968
Signed-off-by: Robert Marko <robimarko@gmail.com>

Shubhamvis98 commented:
> > As @musashino205 suggested, I tried using my real email, but the test still fails. So, is it acceptable to include two Signed-off-by lines — one with my real email and another with the masked one — to satisfy the test requirements?
> 
> Use `git commit --amend --reset-author` (or `git commit --amend --author="First last <e@mail.com>"`) to update the authorship info in your git commit



> > As @musashino205 suggested, I tried using my real email, but the test still fails. So, is it acceptable to include two Signed-off-by lines — one with my real email and another with the masked one — to satisfy the test requirements?
> 
> Use `git commit --amend --reset-author` (or `git commit --amend --author="First last <e@mail.com>"`) to update the authorship info in your git commit



> > As @musashino205 suggested, I tried using my real email, but the test still fails. So, is it acceptable to include two Signed-off-by lines — one with my real email and another with the masked one — to satisfy the test requirements?
> 
> Use `git commit --amend --reset-author` (or `git commit --amend --author="First last <e@mail.com>"`) to update the authorship info in your git commit

Thanks @Rondom

Shubhamvis98 commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq5018-ax830.dts:
```diff
@@ -0,0 +1,487 @@
+		hw_algo = "toggle";
+		hw_margin_ms = <5000>;
+		always-running;
+	};
```
I've updated the MR with kernel defconfig changes.
Thanks

Shubhamvis98 commented on target/linux/qualcommax/config-6.12:
```diff
@@ -194,6 +194,8 @@ CONFIG_GENERIC_TIME_VSYSCALL=y
 CONFIG_GPIOLIB_IRQCHIP=y
 CONFIG_GPIO_CDEV=y
+CONFIG_GPIO_WATCHDOG=y
+# CONFIG_GPIO_WATCHDOG_ARCH_INITCALL is not set
```
The watchdog timer in this board is 5 minutes. So kernel get enough time to initialize the watchdog driver.

robimarko replied:
Ok, makes sense.

robimarko committed:
ipq-wifi: update to Git HEAD (2025-06-23)

63bd0faafd4a Add IPQ5018 BDF for Yuncore AX830.

Link: https://github.com/openwrt/openwrt/pull/18968
Signed-off-by: Robert Marko <robimarko@gmail.com>

Shubhamvis98 commented:
> BTW, you are including ipq-wifi package for this device but it does not exist?

Hi, I've pushed the BDF files on qca-wireless repo. Do I have to make any other changes in openwrt source or somewhere else?

Thanks.

Shubhamvis98 commented:
Hi @robimarko, @musashino205, @georgemoussalem, @Rondom,
Thanks for helping me out with this MR. Really appreciate your support.

Shubhamvis98 commented:
> @Shubhamvis98 Hello, Can you share the original firmware's /sys/firmware/fdt?
> 
> I noticed that the memory layout of different devices of ipq5018 seems to be different. Is the TZ reserved memory of this device 2MB or 4MB?
> 
> The TZ reserve of the mainline kernel is 2MB, I am working on H3C NX48, its TZ reserve memory is 4MB
> 
> This device and H3C NX48 both have 512M memory I think we should check
> 
> The mp series on QCA are all 4M TZ, but there is still a default layout of 2M TZ, I think I should confirm the memory layout
> 
> Thank you

It's 4MB in ax830.

rbqvq commented:
> > @Shubhamvis98 Hello, Can you share the original firmware's /sys/firmware/fdt?
> > I noticed that the memory layout of different devices of ipq5018 seems to be different. Is the TZ reserved memory of this device 2MB or 4MB?
> > The TZ reserve of the mainline kernel is 2MB, I am working on H3C NX48, its TZ reserve memory is 4MB
> > This device and H3C NX48 both have 512M memory I think we should check
> > The mp series on QCA are all 4M TZ, but there is still a default layout of 2M TZ, I think I should confirm the memory layout
> > Thank you
> 
> It's 4MB in ax830.

Thanks, I think there may be something wrong with TZ's reserved memory?

rbqvq commented:
@Shubhamvis98 So can you share the original firmware dtb or fdt?

Shubhamvis98 commented:
@rbqvq, here it is [ax830_fdt](https://gist.github.com/Shubhamvis98/6e3e2a8da8386957a9122bc1b19a1fd9)

sqyexu commented:
> qualcommax: ipq50xx: Add support for Yuncore AX830
> 
> Specifications: SOC: Qualcomm IPQ5018 (64-bit dual-core ARM Cortex-A53 @ 1.0Ghz) Memory: 512MB DDR3L Standard: 802.11ax/ac/b/g/n Flash: SPI NOR 8MB (Winbond W25Q64DW) + NAND 128MB (Winbond W25N01GWZEIG) 2.4G Frequency: 2.4GHz - 2.484GHz 2.4G Wi-Fi standard: 802.11b/g/n/ax 5.8G Frequency: 4.9~5.9G 5.8G Wi-Fi Standard: 802.11 a/n/ac/ax Interface: Optional 1(Without 8081): 1 * 10/100 /1000Mbps RJ45 WAN Port and PoE port; 1* Gigabit Console port; Optional 2(With 8081): 1 * 10/100/1000/2500Mbps RJ45 WAN port and PoE port, 1_10/100/1000Mbps LAN port Buttons: 1 * Reset button, press 10 seconds to revert to default setting Antenna: Build in 4_4dBi dual band MIMO Antenna Data Rate: 3000Mbps （2.4G 600Mbps, 5.8G 2400Mbps） End Users: 128+ 2.4G RF Power: ≤ 23dBm 5.8G RF Power: ≤ 23dBm DC: 12V----2A PoE: 48V (IEEE 802.3at) LED Light: Sys, WAN, LAN Power Consumption: ≤ 20W
> 
> BACKUP YOUR STOCK FIRMWARE:
> 
> * Put openwrt-*-initramfs-uImage.itb to your
>   TFTP server and rename it to initramfs.bin
> * Enable serial console and enter to u-boot cli
>   and exec these commands:
>   ```
>   tftpboot <your_tftp_server_ip>:initramfs.bin
>   bootm
>   ```
> * Once boot completed and you get the openwrt shell
>   execute below commands:
>   ```
>   device=ax830
>   mkdir -p /tmp/fw_backup; cd /tmp/fw_backup
>   rootfs=$(cat /proc/mtd | grep \"rootfs\" | cut -d: -f1)
>   rootfs_1=$(cat /proc/mtd | grep \"rootfs_1\" | cut -d: -f1)
>   dd if=/dev/${rootfs} of=rootfs_${rootfs} bs=1M
>   dd if=/dev/${rootfs_1} of=rootfs_1_${rootfs_1} bs=1M
>   cp /sys/firmware/fdt fdt.dtb
>   md5sum * > md5sum
>   tar -cvzf /tmp/${device}.tar.gz .
>   sum=$(md5sum /tmp/${device}.tar.gz | cut -d' ' -f1)
>   mv /tmp/${device}.tar.gz /tmp/${device}_${sum}.tar.gz
>   echo "stock fw backup saved to: /tmp/${device}_${sum}.tar.gz"
>   ```
> * Upload/save your backup to a safe place.
> 
> STOCK FIRMWARE RECOVERY:
> 
> * Boot initramfs image
> * Upload your backed-up stock fw tarball to the device
>   using scp or download it from the device using wget.
> * Enter device ssh cli or tty and exec:
>   ```
>   cd /tmp && wget <your_web_server_ip>/${stock_fw_backup}.tar.gz`
>   tar -xpzf ${stock_fw_backup}.tar.gz
>   rootfs=$(cat /proc/mtd | grep \"rootfs\" | cut -d: -f1)
>   rootfs_1=$(cat /proc/mtd | grep \"rootfs_1\" | cut -d: -f1)
>   ubiformat /dev/${rootfs} -y -f /tmp/rootfs_${rootfs}
>   ubiformat /dev/${rootfs_1} -y -f /tmp/rootfs_1_${rootfs_1}
>   reboot
>   ```
> 
> INSTALLATION:
> 
> 1. initramfs method
> 
> * Put openwrt-*-initramfs-uImage.itb to your
>   TFTP server and rename it to initramfs.bin
> * Enable serial console and enter to u-boot cli
>   and exec these commands:
>   ```
>   tftpboot <your_tftp_server_ip>:initramfs.bin
>   bootm
>   ```
> * Once boot completed and you get the openwrt shell
>   execute below commands:
>   ```
>   cd /tmp && wget <your_web_server_ip>/factory.ubi`
>   export rootfs=$(cat /proc/mtd | grep rootfs | cut -d: -f1)
>   export rootfs_1=$(cat /proc/mtd | grep rootfs_1 | cut -d: -f1)
>   ubiformat /dev/${rootfs} -y -f factory.ubi
>   ubiformat /dev/${rootfs_1} -y -f factory.ubi
>   reboot
>   ```
> 
> 2. u-boot factory.ubi image method
> 
> * Put factory.ubi to your TFTP server and
>   enter u-boot cli and exec these commands:
>   ```
>   tftpboot <your_tftp_server_ip>:factory.ubi
>   #After downloading is finished:
>   flash rootfs
>   flash rootfs_1
>   reset
>   ```
